src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,275 @@
+
+import edu.washington.escience.myria.DbException;
+import edu.washington.escience.myria.Schema;
+import edu.washington.escience.myria.Type;
+import edu.washington.escience.myria.operator.Operator;
+import edu.washington.escience.myria.operator.UnaryOperator;
+import edu.washington.escience.myria.storage.Tuple;
+import edu.washington.escience.myria.storage.TupleBatch;
+import edu.washington.escience.myria.storage.TupleBuffer;
+import edu.washington.escience.myria.storage.TupleUtils;
+
+/**
+ * This aggregate operator computes the aggregation in streaming manner (requires input sorted on grouping column(s)).
+ * This supports aggregation over multiple columns, with one or more group by columns. Intend to substitute
+ * SingleGroupByAggregate and MultiGroupByAggregate when input is known to be sorted.
+ *
+ * @see Aggregate
+ * @see SingleGroupByAggregate
+ * @see MultiGroupByAggregate
+ */
+public class StreamingAggregate extends UnaryOperator {
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** The schema of the aggregation result. */
+  private Schema aggSchema;
+  /** The schema of the columns indicated by the group keys. */
+  private Schema groupSchema;
+  /** Holds the current grouping key. */
+  private Tuple curGroupKey;
+
+  /** Group fields. **/
+  private final int[] gFields;
+  /** Group field types. **/
+  private final Type[] gTypes;
+  /** An array [0, 1, .., gFields.length-1] used for comparing tuples. */
+  private final int[] gRange;
+  /** Factories to make the Aggregators. **/
+  private final AggregatorFactory[] factories;
+  /** The actual Aggregators. **/
+  private Aggregator[] aggregators;
+  /** The state of the aggregators. */
+  private Object[] aggregatorStates;
+
+  /** Buffer for holding intermediate results. */
+  private transient TupleBuffer resultBuffer;
+  /** Buffer for holding finished results as tuple batches. */
+  private transient ImmutableList<TupleBatch> finalBuffer;
+
+  /**
+   * Groups the input tuples according to the specified grouping fields, then produces the specified aggregates.
+   *
+   * @param child The Operator that is feeding us tuples.
+   * @param gfields The columns over which we are grouping the result.
+   * @param factories The factories that will produce the {@link Aggregator}s for each group.
+   */
+  public StreamingAggregate(@Nullable final Operator child, @Nonnull final int[] gfields,
+      @Nonnull final AggregatorFactory... factories) {
+    super(child);
+    gFields = Objects.requireNonNull(gfields, "gfields");
+    gTypes = new Type[gfields.length];
+    this.factories = Objects.requireNonNull(factories, "factories");
+    Preconditions.checkArgument(gfields.length > 0, " must have at least one group by field");
+    Preconditions.checkArgument(factories.length > 0, "to use StreamingAggregate, must specify some aggregates");
+    gRange = new int[gfields.length];
+    for (int i = 0; i < gfields.length; ++i) {
```

Comment:
Nitpick: I'd prefer `gRange.length` here since the array bounds should be those of the array being modified.